sympy/physics/mechanics/tests/test_joint.py

```diff
@@ -0,0 +1,186 @@
+    assert A.ang_vel_in(N) == omega * N.y
+    assert A.ang_vel_in(B) == - alpha * A.z
+    assert N.ang_vel_in(B) == - omega * N.y - alpha * A.z
+
+    #Check kde
+    assert J1.kdes() == [omega - theta.diff(t)]
+    assert J2.kdes() == [alpha - phi.diff(t)]
+
+    #Check pos of masscenters
+    assert C.masscenter.pos_from(rod.masscenter) == lA*A.z
+    assert rod.masscenter.pos_from(plate.masscenter) == - lC * A.z
+
+    #Check Linear Velocities
+    assert rod.masscenter.vel(N) == (h/4 - lB/2)*omega*A.x
+    assert plate.masscenter.vel(N) == ((h/4 - lB/2)*omega + (h/4 + lB/2)*omega)*A.x
+
+def test_pinjoint_arbitrary_axis():
+    def generate_body():
+        N = ReferenceFrame('N')
+        A = ReferenceFrame('A')
+        P = Body('P', frame=N)
+        C = Body('C', frame=A)
+        return N, A, P, C
+
+    theta, omega = dynamicsymbols('J_theta, J_omega')
+
+    #When the bodies are attached though masscenters but axess are opposite.
+    N, A, P, C = generate_body()
+    PinJoint('J', P, C, child_axis=-A.x)
+
+    assert -A.x.angle_between(N.x) == 0
+    assert A.dcm(N) == Matrix([[1, 0, 0], [0, cos(theta), sin(theta)], [0, -sin(theta), cos(theta)]])
+    assert A.ang_vel_in(N) == omega*N.x
+    assert C.masscenter.pos_from(P.masscenter) == 0
+    assert C.masscenter.vel(N) == 0
+
+    #When axes are different and parent joint is at masscenter but child joint is at a unit vector from
+    #child masscenter.
+    N, A, P, C = generate_body()
+    PinJoint('J', P, C, child_axis=A.y, child_joint_pos=A.x)
+
+    assert A.y.angle_between(N.x) == 0 #Axis are aligned
+    assert A.dcm(N) == Matrix([[0, -cos(theta), -sin(theta)], [1, 0, 0], [0, -sin(theta), cos(theta)]])
+    assert A.ang_vel_in(N) == omega*N.x
+    assert A.ang_vel_in(N).express(A) == omega * A.y
+    assert C.masscenter.vel(N) == omega*A.z
+    assert C.masscenter.pos_from(P.masscenter) == -A.x
+
+    #Similar to previous case but wrt parent body
+    N, A, P, C = generate_body()
+    PinJoint('J', P, C, parent_axis=N.y, parent_joint_pos=N.x)
+
+    assert N.y.angle_between(A.x) == 0 #Axis are aligned
+    assert A.dcm(N) == Matrix([[0, 1, 0], [-cos(theta), 0, sin(theta)], [sin(theta), 0, cos(theta)]])
+    assert A.ang_vel_in(N) == omega*N.y
+    assert A.ang_vel_in(N).express(A) == omega* A.x
+    assert C.masscenter.vel(N).simplify() == - omega*N.z
+    assert C.masscenter.pos_from(P.masscenter) == N.x
+
+    #Both joint pos id defined but different axes
+    N, A, P, C = generate_body()
+    PinJoint('J', P, C, parent_joint_pos=N.x, child_joint_pos=A.x, child_axis=A.x+A.y)
+    assert N.x.angle_between(A.x + A.y).simplify() == 0 #Axis are aligned
+    assert A.dcm(N).simplify() == Matrix([[sqrt(2)/2, -sqrt(2)*cos(theta)/2, -sqrt(2)*sin(theta)/2], \
+        [sqrt(2)/2, sqrt(2)*cos(theta)/2, sqrt(2)*sin(theta)/2], [0, -sin(theta), cos(theta)]])
+    assert A.ang_vel_in(N) == omega*N.x
+    assert A.ang_vel_in(N).express(A).simplify() == (omega*A.x + omega*A.y)/sqrt(2)
```

Comment:
`A.ang_vel_in(N).angle_between(A.x + A.y) == 0`

Reply:
`A.ang_vel_in(N).magnitude() == omega`

Reply:
> A.ang_vel_in(N).angle_between(A.x + A.y)

Isn't this test same as `assert N.x.angle_between(A.x + A.y).simplify() == 0 #Axis are aligned` this one

Reply:
No, it isn't.

Reply:
@moorepants How can I simplify `sqrt(omega**2)` where `omega` is a dynamicsymbol to `omega`? `.simplify()` isn't working.

Reply:
The key things to check for the velocities are:

1. The angular velocity should be aligned with the pin joint axis expressed in either frame. There is the matter of positive or negative depending on the sense of N_omega_A = - A_omega_N.
2. The linear velocity of the mass centers should be normal to the position vector connecting the pin joint point and the mass center.

Both of those should hold in the arbitrary case. They both ensure the direction of the velocity vectors are correct (at least partly, there is still magnitude and sense to check).

Reply:
> @moorepants How can I simplify sqrt(omega**2) where omega is a dynamicsymbol to omega? .simplify() isn't working.


This is because omega is by default a complex number.

Reply:
> This is because omega is by default a complex number.

Okay, got it.